### PR TITLE
Assignement : Amazon cart functionality - Web

### PR DESCRIPTION
**Pull request by Arunpradeep-Sakthivel**

### CODE REVIEW

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+public class AmazonGridWallPageWeb extends AmazonGridWallPageScreen {
+    private final Visual visually;
+    private final Driver driver;
+    public static final String SCREEN_NAME = AmazonGridWallPageWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+    private final By header_Results = By.xpath("//span[text()='RESULTS']");
+    private final By list_SearchResults = By.xpath("//div[@data-component-type='s-search-result']");
+
+    public By productFromTheSearchResults(String productToBeSelected) {
+       By productFromTheSearchResults = By.xpath("//div[@data-component-type='s-search-result'][1]//h2//a//span[text()='"+productToBeSelected+"']");
```
Comment (ShamaUgale) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java:21]: write better xpaths.. indexes should not be used.

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                .getId();
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = SAMPLE_TEST_CONTEXT.ME;
+        this.currentPlatform = Runner.platform;
+    }
+
+    public CartPageBL clickCartButton() {
+        AmazonCartPageScreen.get().clickCartButton();
+        return this;
```
Comment (ShamaUgale) [inline src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java:45]: assertion missing

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.screen.amazon.AmazonCartPageScreen;
+import com.znsio.sample.e2e.screen.amazon.AmazonGridWallPageScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CartPageBL {
+    private static final Logger LOGGER = Logger.getLogger(LandingBL.class.getName());
```
Comment (ShamaUgale) [inline src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java:16]: LandingBL.class.getName()?

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonCartPageWeb.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    public AmazonCartPageWeb(Driver driver, Visual visually) {
+        this.driver = driver;
+        this.visually = visually;
+    }
+
+    @Override
+    public CartPageBL clickCartButton() {
+        driver.waitForClickabilityOf(button_Cart,20).click();
+        visually.takeScreenshot(SCREEN_NAME, "After Selecting the cart button in Express checkout path page");
+        return new CartPageBL();
```
Comment (ShamaUgale) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonCartPageWeb.java:31]: why screen is returning BL objects?

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java (hunk 3 above)
Comment (poojamanna18) [inline src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java:16]: Update the class name

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/GridWallPageBL.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import com.znsio.sample.e2e.businessLayer.jiomeet.LandingBL;
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.screen.amazon.AmazonGridWallPageScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GridWallPageBL {
+    private static final Logger LOGGER = Logger.getLogger(LandingBL.class.getName());
```
Comment (poojamanna18) [inline src/test/java/com/znsio/sample/e2e/businessLayer/amazon/GridWallPageBL.java:15]: update the class name

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.screen.amazon.AmazonGridWallPageScreen;
+import com.znsio.sample.e2e.screen.amazon.AmazonProductPageScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ProductPageBL {
+    private static final Logger LOGGER = Logger.getLogger(HomePageBL.class.getName());
```
Comment (poojamanna18) [inline src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java:15]: update the class name

**Review by poojamanna18 — COMMENTED** [review]
Review comments

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java (hunk 1 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java]: Updated the code

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java (hunk 2 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java:44]: Added assertion

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java (hunk 3 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java]: Updated the class name correctly

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonCartPageWeb.java (hunk 4 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonCartPageWeb.java]: Updated the return type

File: src/test/resources/com/znsio/sample/e2e/features/amazon.feature [hunk 7]
```diff
@@ -0,0 +1,14 @@
+@amazon
+  Feature: Amazon
+  @web @prod
+  Scenario: validate add to cart functionality for unregistered user
+    Given User is on amazon homepage
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/resources/com/znsio/sample/e2e/features/amazon.feature]: Updated the feature file

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/HomePageBL.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = userPersona;
+        this.currentPlatform = forPlatform;
+        Runner.setCurrentDriverForUser(userPersona, forPlatform, context);
+    }
+
+    public GridWallPageBL searchProduct(String productToSearch)  {
+        AmazonHomePageScreen.get().searchProduct(productToSearch);
+        return new GridWallPageBL();
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/HomePageBL.java:31]: Added assertion

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    public ProductPageBL() {
+        long threadId = Thread.currentThread()
+                .getId();
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = SAMPLE_TEST_CONTEXT.ME;
+        this.currentPlatform = Runner.platform;
+    }
+    public ProductPageBL verifyProductPageAndProductDesc(String expectedProductDesc){
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java]: Changed the method name

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    public ProductPageBL verifyAddToCartButton() {
+        Boolean isProductPageDisplayed = AmazonProductPageScreen.get().verifyAddToCartButton();
+        assertThat(isProductPageDisplayed).as(currentUserPersona + "Add to cart button is not displayed in product page")
+                .isTrue();
+        return this;
+    }
+
+    public CartPageBL clickAddToCartButton() {
+        AmazonProductPageScreen.get().clickAddToCartButton();
+        return new CartPageBL();
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java:58]: Updated

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java]: Updated the code

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        return isResultsSizeIsMoreThanOne;
+    }
+
+    @Override
+    public ProductPageBL selectProduct(String productToBeSelected) {
+        LOGGER.info(String.format("ProductDesc: "+productToBeSelected+""));
+        driver.waitTillElementIsPresent(productFromTheSearchResults(productToBeSelected),10).click();
+        visually.takeScreenshot(SCREEN_NAME, "After Selecting the product from the serach results");
+        return new ProductPageBL();
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonGridWallPageWeb.java]: Updated

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonHomePageWeb.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    public GridWallPageBL searchProduct(String productToSeacrh) {
+        visually.takeScreenshot(SCREEN_NAME, "Before entering the product name in serach bar");
+        LOGGER.info(String.format("ProductName: "+productToSeacrh+""));
+        WebElement enterProduct = driver.waitTillElementIsPresent(searchBar, 20);
+        enterProduct.sendKeys(productToSeacrh);
+        visually.takeScreenshot(SCREEN_NAME, "Before selecting the product from the Search bar suggestions");
+        driver.waitTillElementIsPresent(selectSearchResult(productToSeacrh),5).click();
+        visually.takeScreenshot(SCREEN_NAME, "After selecting the product from the Search bar suggestions");
+        return new GridWallPageBL();
```
[the remark this answers is not in the document]
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/AmazonHomePageWeb.java]: Updated

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java (hunk 3 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/CartPageBL.java]: Updated

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/GridWallPageBL.java (hunk 5 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/GridWallPageBL.java]: Updated

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java (hunk 6 above)
Comment (Arunpradeep-Sakthivel) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductPageBL.java]: Updated

**Pull request closed without merging**